### ISSUE TITLE
STACK parser and tellsimp

### ISSUE THREAD

**Issue by sangwinc** [maths/moodle-qtype_stack#687]
The STACK parser does not like the output from tellsimp/tellsimpafter.  To recreate the problem use the following as variables

```
matchdeclare(pmpatex1,true);
matchdeclare(pmpatex2,true);
tellsimpafter(-(pmpatex1 #pm# pmpatex2),(-pmpatex1) #pm# pmpatex2);
```

The problem is that the output from tellsimpafter is `[*rule1, simptimes]` which has an awkward `*` symbol.

**Comment by aharjula** [comment]
Indeed and if we parse it there we would have to parse it everywhere and that just cannot be. In my opinion, all commands that have this type of behaviour that has no sensible return value to be used should be added to the session as silent ones. Basically, if you are already extracting those context vars or block-externals and the like you could simply make them silent.

Though you should note that in the cache-branch this should already be fine as nothing returns anything outside CASText and PRT/input-processing. The only places where things break are those debug outputs on various pages, you would just need to silence those things at that level...

Overall, I have been happy that the steps started at 4.3 and that came from the cache-branch and not in the castext2-branch are getting to the point where we no longer need to worry about serialisation (into maxima input command syntax) of the return values as we are just returning the end results of PRTs and CASText rendering or input validation all of which are inherently serialisable no matter what went into them. As soon as we can get to the point where the intermediate results inside the CAS no longer need to be serialisable we can finally start thinking about how to allow people to use that graph theory library and others that have such interesting features (as long as they do not do everything with files).

Of course, I understand that for debugging purposes it is nice to be able to dump out the intermediate results and we should always aim to provide that feature as long as possible. However, we can do so much more if we don't limit ourselves too much. To me, it would make sense to start tagging the security-map with functions that generate results that are not serialisable and simply check the statements for such functions when generating any debug output, silencing them and just writing in the debug that the value of this statement is not something we can present.

But no the parser will not parse that, and unless you want to switch parsers of the output based on the input command it will never do that. Note, that not all CAS-evaluatables do automatic-parsing of the return value, in addition to the silent ones there are variants that receive the raw string form and can do simplified parsing of their own, like the [CASText2-evaluatable](https://github.com/maths/moodle-qtype_stack/blob/castext2/stack/cas/castext2/castext2_evaluatable.class.php#L33) that uses a  simplified parser to parse the nested list of strings it receives and can skip all the operator parsing required for the full-featured operation. For debug purposes, one could just as well ask for the raw value and skip the parsing entirely...

**Comment by sangwinc** [comment]
Ok Matti, but how do I make them silent?!  I think tellsimp and tellsimpafter should always be silent.

**Comment by aharjula** [comment]
Remember how I have often commented about the use of AST-container vs. its silent parent-class inside the keyval-class. What you want to do is that when generating the AST-containers to evaluate for the session you check if the AST contains something that would not output correctly and switch to a container that does not output stuff, or outputs it in some other form. Basically, you recently separated the context vars there, when doing so you could have just as well made them use silent containers at the same time.

You do not tell Maxima to be silent as you can't use `$` but instead, you tell the cassession2 to not collect the values related to that statement by using a cas_evaluatable (e.g. the AST-container) of the type that does not ask for them to be collected.

Basically, do the same you did with the context-vars etc. but instead of reordering the statements convert the matching containers to silent containers.
